File: 5Days-SoftwareArchitecture/Java/Day-1 CC Metrics/RfcDemo.java
import java.util.Random;

class AuthService {
    boolean checkCredentials(String user) { return true; }
}

class Database {
    void connect() { System.out.println("[DB] connect"); }
    void save(String payload) { System.out.println("[DB] save: " + payload); }
    void close() { System.out.println("[DB] close"); }
}

class EmailService {
    void sendReceipt(String to, String body) { System.out.println("[Email] to=" + to); }
}

class Logger {
    void info(String m) { System.out.println("[Log] " + m); }
    void error(String m) { System.out.println("[Log-Err] " + m); }
}

class Order {
    final String id; final double amount;
    Order(String id, double amount){ this.id = id; this.amount = amount; }
    public String toString(){ return id + ":" + amount; }
}

class OrderProcessor { // POOR: chatty object (high RFC)
    private final AuthService auth = new AuthService();
    private final Database db = new Database();
    private final EmailService email = new EmailService();
    private final Logger logger = new Logger();

    void process(Order o, String user) {
        logger.info("Start " + o);
        if (!auth.checkCredentials(user)) { logger.error("auth fail"); return; }
        db.connect();
        db.save(o.toString());
        email.sendReceipt(user, "Thanks for order " + o.id);
        logger.info("Done " + o);
        db.close();
    }

    // illustrative RFC count (not computed via analysis) 
    static int illustrativeRFC() {
        int M = 1; // methods owned (process)
        int R = 6; // external methods called: auth.checkCredentials, db.connect, db.save, email.sendReceipt, logger.info, db.close
        return M + R;
    }
}

class TransactionFacade { // Facade that groups many calls
    private final AuthService auth;
    private final Database db;
    private final EmailService email;
    private final Logger logger;

    TransactionFacade(AuthService auth, Database db, EmailService email, Logger logger) {
        this.auth = auth; this.db = db; this.email = email; this.logger = logger;
    }

    boolean handleTransaction(Order o, String user) {
        logger.info("Facade start " + o);
        if (!auth.checkCredentials(user)) { logger.error("auth fail"); return false; }
        db.connect();
        db.save(o.toString());
        email.sendReceipt(user, "Thanks for order " + o.id);
        logger.info("Facade done " + o);
        db.close();
        return true;
    }
}

class CleanProcessor { // GOOD: low RFC because it delegates to a single facade call
    private final TransactionFacade facade;
    CleanProcessor(TransactionFacade facade){ this.facade = facade; }

    void process(Order o, String user){
        facade.handleTransaction(o, user);
    }

    static int illustrativeRFC(){
        int M = 1; // process
        int R = 1; // facade.handleTransaction
        return M + R;
    }
}

public class RfcDemo {
    public static void main(String[] args){
        Order o1 = new Order("o-101", 49.99);
        Order o2 = new Order("o-202", 149.50);

        System.out.println("=== POOR: High RFC (chatty) ===");
        OrderProcessor p = new OrderProcessor();
        p.process(o1, "alice@example.com");
        System.out.println("OrderProcessor RFC (illustrative) = " + OrderProcessor.illustrativeRFC());

        System.out.println("\n=== GOOD: Low RFC (facade) ===");
        TransactionFacade facade = new TransactionFacade(new AuthService(), new Database(), new EmailService(), new Logger());
        CleanProcessor cp = new CleanProcessor(facade);
        cp.process(o2, "bob@example.com");
        System.out.println("CleanProcessor RFC (illustrative) = " + CleanProcessor.illustrativeRFC());

        System.out.println("\n=== Creative: Run many mixed processors to show behavior ===");
        Random rnd = new Random();
        for (int i=0;i<3;i++){
            Order o = new Order("batch-"+i, rnd.nextDouble()*200);
            if (rnd.nextBoolean()) p.process(o, "user@x");
            else cp.process(o, "user@y");
        }
    }
}

File: 5Days-SoftwareArchitecture/typesript/Day1- Metrics/rfcDemo.js
var AuthService = /** @class */ (function () {
    function AuthService() {
    }
    AuthService.prototype.checkCredentials = function (user) { return true; };
    return AuthService;
}());
var Database = /** @class */ (function () {
    function Database() {
    }
    Database.prototype.connect = function () { console.log("[DB] connect"); };
    Database.prototype.save = function (payload) { console.log("[DB] save:", payload); };
    Database.prototype.close = function () { console.log("[DB] close"); };
    return Database;
}());
var EmailService = /** @class */ (function () {
    function EmailService() {
    }
    EmailService.prototype.sendReceipt = function (to, body) { console.log("[Email] to=", to); };
    return EmailService;
}());
var Logger = /** @class */ (function () {
    function Logger() {
    }
    Logger.prototype.info = function (m) { console.log("[Log]", m); };
    Logger.prototype.error = function (m) { console.log("[Log-Err]", m); };
    return Logger;
}());
var OrderProcessor = /** @class */ (function () {
    function OrderProcessor() {
        this.auth = new AuthService();
        this.db = new Database();
        this.email = new EmailService();
        this.logger = new Logger();
    }
    OrderProcessor.prototype.process = function (o, user) {
        this.logger.info("Start " + JSON.stringify(o));
        if (!this.auth.checkCredentials(user)) {
            this.logger.error("auth fail");
            return;
        }
        this.db.connect();
        this.db.save(JSON.stringify(o));
        this.email.sendReceipt(user, "Thanks " + o.id);
        this.logger.info("Done " + JSON.stringify(o));
        this.db.close();
    };
    OrderProcessor.illustrativeRFC = function () {
        var M = 1;
        var R = 6;
        return M + R;
    };
    return OrderProcessor;
}());
var TransactionFacade = /** @class */ (function () {
    function TransactionFacade(auth, db, email, logger) {
        this.auth = auth;
        this.db = db;
        this.email = email;
        this.logger = logger;
    }
    TransactionFacade.prototype.handleTransaction = function (o, user) {
        this.logger.info("Facade start " + JSON.stringify(o));
        if (!this.auth.checkCredentials(user)) {
            this.logger.error("auth fail");
            return false;
        }
        this.db.connect();
        this.db.save(JSON.stringify(o));
        this.email.sendReceipt(user, "Thanks " + o.id);
        this.logger.info("Facade done " + JSON.stringify(o));
        this.db.close();
        return true;
    };
    return TransactionFacade;
}());
var CleanProcessor = /** @class */ (function () {
    function CleanProcessor(facade) {
        this.facade = facade;
    }
    CleanProcessor.prototype.process = function (o, user) { this.facade.handleTransaction(o, user); };
    CleanProcessor.illustrativeRFC = function () { return 1 + 1; };
    return CleanProcessor;
}());
console.log("=== POOR: High RFC (chatty) ===");
var poor = new OrderProcessor();
poor.process({ id: "o-11", amount: 9.99 }, "alice@x.com");
console.log("OrderProcessor RFC (illustrative) =", OrderProcessor.illustrativeRFC());
console.log("\n=== GOOD: Low RFC (facade) ===");
var facade = new TransactionFacade(new AuthService(), new Database(), new EmailService(), new Logger());
var clean = new CleanProcessor(facade);
clean.process({ id: "o-22", amount: 29.99 }, "bob@y.com");
console.log("CleanProcessor RFC (illustrative) =", CleanProcessor.illustrativeRFC());
console.log("\n=== Creative: batch run ===");
for (var i = 0; i < 3; i++) {
    var o = { id: "batch-".concat(i), amount: Math.random() * 300 };
    if (Math.random() > 0.5)
        poor.process(o, "u@x");
    else
        clean.process(o, "u@y");
}
